engine/time/src/main/java/io/deephaven/time/calendar/FastConcurrentCache.java

```diff
@@ -0,0 +1,130 @@
+//
+// Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+//
+package io.deephaven.time.calendar;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+/**
+ * A cache that is designed to be fast when accessed concurrently. When created, the cache uses a ConcurrentHashMap to
+ * store values. When the fast cache is enabled, the cache is converted to a HashMap. This is done because HashMap is
+ * faster than {@link ConcurrentHashMap} when accessed concurrently. The fast cache is immutable, so it is safe to use
+ * the HashMap for optimal concurrent access.
+ * <p>
+ * The fast cache population happens on a separate thread. This is done to avoid blocking the calling thread. The
+ * calling thread can wait for the fast cache to be populated by calling {@link #enableFastCache} with the wait
+ * parameter set to true. This will block the calling thread until the fast cache is populated.
+ *
+ * @param <K> the key type
+ * @param <V> the value type
+ */
+class FastConcurrentCache<K, V> {
+
+    private final Function<K, V> valueComputer;
+    private final Map<K, V> concurrentCache = new ConcurrentHashMap<>();
+    private final Map<K, V> fastCache = new HashMap<>();
```

Comment:
I'm very skeptical that we want to add the complexity of a two-tiered structure like this with bespoke setup and limitations on get.

Instead, we might take the approach of using StampedLock in read vs write mode. I've done a quick PoC w/ HashMap and StampedLock, and the results appear to be similar or better.

```java
    public V get(K key) {
        long stamp = lock.readLock();
        try {
            {
                final V existing = map.get(key);
                if (existing != null) {
                    return existing;
                }
            }
            final V newValue = valueComputer.apply(key);
            stamp = toWriteLock(stamp);
            final V existing = map.putIfAbsent(key, newValue);
            return existing == null ? newValue : existing;
        } finally {
            lock.unlock(stamp);
        }
    }

    private long toWriteLock(final long readStamp) {
        final long ws = lock.tryConvertToWriteLock(readStamp);
        if (ws != 0) {
            return ws;
        }
        lock.unlockRead(readStamp);
        return lock.writeLock();
    }
```

This also does away with the need for "enabled vs disabled" distinction. There may be advantage to "preloading" keys if desired, but I haven't tested thoroughly.

Reply:
The complex structure was the result of trying to make both Ryan and Charles happy with perf tradeoffs.  I'll give this a try to compare.

Reply:
Performance using the proposed `StampedLock` design.  Caches are prepopulated to easily compare against the existing implementation.

```
// 10M - StampedLock
TIME: calendarDay-StampedLock (Table): 474 ms.      <<< vs 416 ms
TIME: calendarDay-StampedLock (PTable): 2982 ms   <<< vs 503 ms

TIME: diffBusinessYears-StampedLock (Table): 19523 ms.        <<< vs 9720 ms
TIME: diffBusinessYears-StampedLock (PTable): 284067 ms    <<< vs 1757 ms
```

As was seen in other locking designs that I attempted, partitioned table perf is bad whenever a lock is introduced.

Possibly an optimistic read lock gives better results.  Trying that next.

Reply:
More experiments.  In each of these cases, the cache is prepopulated to make comparisons easy.

```
// 10M - "complex" two tiered cache

TIME: calendarDay-NEW_FAST (Table): 419 ms
TIME: calendarDay-NEW_FAST (PTable): 616 ms

TIME: diffBusinessYears-NEW_FAST (Table): 9785 ms
TIME: diffBusinessYears-NEW_FAST (PTable): 1762 ms

// 10M - StampedLock using readLock

TIME: calendarDay-NEW_FAST (Table): 386 ms
TIME: calendarDay-NEW_FAST (PTable): 2658 ms

TIME: diffBusinessYears-NEW_FAST (Table): 18645 ms
TIME: diffBusinessYears-NEW_FAST (PTable): 249826 ms

// 10M - StampledLock using readOptimisticLock

TIME: calendarDay-NEW_FAST (Table): 449 ms
TIME: calendarDay-NEW_FAST (PTable): 603 ms

TIME: diffBusinessYears-NEW_FAST (Table): 13625 ms
TIME: diffBusinessYears-NEW_FAST (PTable): 2179 ms
```

Here the `StampledLock.readLock` performance is unacceptable.  `StampedLock.readOptimisticLock` is much more in line with what we need to see.  For the simple `calendarDay` case, it is in line with the two-tier cache.  For the more complex `diffBusinessYears` case, it appears to be 40% slower for tables and 23% slower for partitioned tables.

A second run produced similar results:
```
// 10M - "complex" two tiered cache

TIME: calendarDay-NEW_FAST (Table): 356 ms
TIME: calendarDay-NEW_FAST (PTable): 465 ms

TIME: diffBusinessYears-NEW_FAST (Table): 9982 ms
TIME: diffBusinessYears-NEW_FAST (PTable): 1680 ms

// 10M - StampledLock using readOptimisticLock

TIME: calendarDay-NEW_FAST (Table): 536 ms
TIME: calendarDay-NEW_FAST (PTable): 560 ms

TIME: diffBusinessYears-NEW_FAST (Table): 13480 ms
TIME: diffBusinessYears-NEW_FAST (PTable): 2240 ms
```

Reply:
There is clearly a tradeoff with slower execution and less memory being used because the entire cache does not need to be populated.

Reply:
So, I'm still interested in trying to get rid of the two-tiered system; how do performance numbers look in the "normal" case for the two-tiered system (ie, no fast-cache / no pre-population) vs the "normal" case for the read optimistic system (ie, no pre-population).

Reply:
Below are performance numbers with and without fast-cache and pre-population.

With no-fast-cache / no-prepopulation, the two-tier case reduces to a ConcurrentHashMap.  In this case, the data supports readOptimisticLock over the ConcurrentHashMap by 4-9%

With fast-cache / prepopulation, the two-tier case reduces to a HashMap.  In this case, the data strongly supports the two-tier case by about 40%.

For max speed, a two-tier cache with readOptimisticLock in place of ConcurrentHashMap would probably be the fastest but would be yet more complex.

# Case 1: no fast-cache and no pre-population. (two-tier => ConcurrentHashMap)

There is quite a bit of noise in the 10M row experiments:

Run 1:
```
// 10M - "complex" two tiered cache

TIME: calendarDay-NEW (Table): 402 ms
TIME: calendarDay-NEW (PTable): 591 ms

TIME: diffBusinessYears-NEW (Table): 13674 ms
TIME: diffBusinessYears-NEW (PTable): 2433 ms

// 10M - StampledLock using readOptimisticLock

TIME: calendarDay-NEW (Table): 694 ms
TIME: calendarDay-NEW (PTable): 828 ms

TIME: diffBusinessYears-NEW (Table): 13699 ms
TIME: diffBusinessYears-NEW (PTable): 2268 ms
```

Run 2:
```
// 10M - "complex" two tiered cache

TIME: calendarDay-NEW (Table): 593 ms
TIME: calendarDay-NEW (PTable): 864 ms

TIME: diffBusinessYears-NEW (Table): 14354 ms
TIME: diffBusinessYears-NEW (PTable): 2422 ms

// 10M - StampledLock using readOptimisticLock

TIME: calendarDay-NEW (Table): 468 ms
TIME: calendarDay-NEW (PTable): 475 ms

TIME: diffBusinessYears-NEW (Table): 14487 ms
TIME: diffBusinessYears-NEW (PTable): 2348 ms
```

Going to a larger size:
```
// 100M - "complex" two tiered cache

TIME: calendarDay-NEW (Table): 6814 ms
TIME: calendarDay-NEW (PTable): 5314 ms

TIME: diffBusinessYears-NEW (Table): 143081 ms
TIME: diffBusinessYears-NEW (PTable): 24376 ms

// 100M - StampledLock using readOptimisticLock

TIME: calendarDay-NEW (Table): 3726 ms
TIME: calendarDay-NEW (PTable): 5451 ms

TIME: diffBusinessYears-NEW (Table): 137975 ms
TIME: diffBusinessYears-NEW (PTable): 22433 ms
```

# Case 2: fast-cache and pre-population. (two-tier => HashMap)

```
// 100M - "complex" two-tiered cache

TIME: calendarDay-NEW_FAST (Table): 3375 ms
TIME: calendarDay-NEW_FAST (PTable): 4998 ms

TIME: diffBusinessYears-NEW_FAST (Table): 106025 ms
TIME: diffBusinessYears-NEW_FAST (PTable): 18543 ms

// 100M - StampledLock using readOptimisticLock

TIME: calendarDay-NEW_FAST (Table): 3734 ms
TIME: calendarDay-NEW_FAST (PTable): 4998 ms

TIME: diffBusinessYears-NEW_FAST (Table): 146659 ms
TIME: diffBusinessYears-NEW_FAST (PTable): 25245 ms
```

Reply:
```
// 100M - current prod impl

TIME: calendarDay-OLD (Table): 6166 ms
TIME: calendarDay-OLD (PTable): 8659 ms

TIME: diffBusinessYears-OLD (Table): DNF
TIME: diffBusinessYears-OLD (PTable): DNF
```

Reply:
Timings from the latest cleaned up StampedLock readOptimisticLock with no cache population.
```
// 100M - StampledLock using readOptimisticLock no cache population

TIME: calendarDay-NEW (Table): 5499 ms
TIME: calendarDay-NEW (PTable): 6095 ms

TIME: diffBusinessYears-NEW (Table): 118555 ms
TIME: diffBusinessYears-NEW (PTable): 19881 ms
```

Reply:
No fast cache.  No cache population.
```
// 100M - KeyedObjectHash
TIME: calendarDay-NEW (Table): 4789 ms
TIME: calendarDay-NEW (PTable): 5986 ms

TIME: diffBusinessYears-NEW (Table): 162804 ms
TIME: diffBusinessYears-NEW (PTable): 27391 ms

// 100M - KeyedIntObjectHash
TIME: calendarDay-NEW (Table): 3287 ms
TIME: calendarDay-NEW (PTable): 5220 ms

TIME: diffBusinessYears-NEW (Table): 109621 ms
TIME: diffBusinessYears-NEW (PTable): 20734 ms
```

Reply:
```
// 100M - KeyedIntObjectHash with a few more improvements

TIME: calendarDay-NEW (Table): 4550 ms
TIME: calendarDay-NEW (PTable): 5823 ms

TIME: diffBusinessYears-NEW (Table): 102374 ms
TIME: diffBusinessYears-NEW (PTable): 17374 ms
```